### PR TITLE
Update/unified video player

### PR DESCRIPTION
**Pull request by devgabrielcoman**
## JIRA
KWS-5180](https://jira.superawesome.tv/browse/KWS-5180)

## Description
a) Rewritten the video part in Swift
b) Rewritten or amended the `SAVideoAd` class; a class full of static methods; it's simpler now; 
c) Split out eventing for video into `SAVideoEvents` and `SAVideoClick`
d) The activity that plays the video now only gets injected with some config objects (an `SAAd` instance, an `SAEvents` instance, etc) and builds internally what it needs (`video player`, `media control` ,etc)

a + b + c ==> less shared state between objects, less head-aches for me.

### CODE REVIEW

File: Example/SuperAwesome/AdMobController.m [hunk 1]
```diff
@@ -1,12 +1,12 @@
 #import "AdMobController.h"
-#import <GoogleMobileAds/GoogleMobileAds.h>
-#import "AwesomeAds.h"
-#import "SAAdMobExtras.h"
+//#import <GoogleMobileAds/GoogleMobileAds.h>
+//#import "AwesomeAds.h"
+//#import "SAAdMobExtras.h"
```
Comment (Tbruni85) [inline Example/SuperAwesome/AdMobController.m]: can these be removed?

File: Example/SuperAwesome/AdMobController.m [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
 
-@interface AdMobController () <GADBannerViewDelegate, GADInterstitialDelegate, GADRewardBasedVideoAdDelegate>
+@interface AdMobController () /*<GADBannerViewDelegate, GADInterstitialDelegate, GADRewardBasedVideoAdDelegate>*/
```
Comment (Tbruni85) [inline Example/SuperAwesome/AdMobController.m]: these too?

File: Example/SuperAwesome/AdMobController.m [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
 
-@property (weak, nonatomic) IBOutlet GADBannerView *banner;
-@property (nonatomic, strong) GADInterstitial *interstitial;
+//@property (weak, nonatomic) IBOutlet GADBannerView *banner;
+//@property (nonatomic, strong) GADInterstitial *interstitial;
```
Comment (Tbruni85) [inline Example/SuperAwesome/AdMobController.m]: these comments?

File: Example/SuperAwesome/AdMobController.m [hunk 4]
```diff
@@ -15,49 +15,49 @@ @implementation AdMobController
...
+//    extra.closeAtEndEnabled = true;
+//    extra.closeButtonEnabled = false;
+//    extra.parentalGateEnabled = false;
+//    extra.smallCLickEnabled = true;
+//    extra.configuration = STAGING;
+//    extra.orientation = LANDSCAPE;
+//    [vidRequest registerAdNetworkExtras:extra];
+//    [[GADRewardBasedVideoAd sharedInstance] loadRequest:vidRequest
+//                                           withAdUnitID:@"ca-app-pub-7706302691807937/9233253206"];
+//    [[GADRewardBasedVideoAd sharedInstance] setDelegate:self];
```
Comment (Tbruni85) [inline Example/SuperAwesome/AdMobController.m]: remove these if not needed please

File: Example/SuperAwesome/AdMobController.m [hunk 5]
```diff
@@ -66,115 +66,115 @@ - (void)viewDidLoad {
...
-    if (self.interstitial.isReady) {
-        [self.interstitial presentFromRootViewController:self];
-    } else {
-        NSLog(@"[SuperAwesome | AdMob] Interstitial Ad wasn't ready");
-    }
+//    if (self.interstitial.isReady) {
+//        [self.interstitial presentFromRootViewController:self];
+//    } else {
+//        NSLog(@"[SuperAwesome | AdMob] Interstitial Ad wasn't ready");
+//    }
```
Comment (Tbruni85) [inline Example/SuperAwesome/AdMobController.m]: remove?

File: Example/SuperAwesome/AdMobController.m [hunk 6]
```diff
@@ -66,115 +66,115 @@ - (void)viewDidLoad {
...
-    if ([[GADRewardBasedVideoAd sharedInstance] isReady]) {
-        [[GADRewardBasedVideoAd sharedInstance] presentFromRootViewController:self];
-    }else {
-        NSLog(@"[SuperAwesome | AdMob] Video ad wasn't ready");
-    }
+//    if ([[GADRewardBasedVideoAd sharedInstance] isReady]) {
+//        [[GADRewardBasedVideoAd sharedInstance] presentFromRootViewController:self];
+//    }else {
+//        NSLog(@"[SuperAwesome | AdMob] Video ad wasn't ready");
+//    }
```
Comment (Tbruni85) [inline Example/SuperAwesome/AdMobController.m]: remove?

File: Example/SuperAwesome/AdMobController.m [hunk 7]
```diff
@@ -66,115 +66,115 @@ - (void)viewDidLoad {
...
+//    NSLog(@"[SuperAwesome | AdMob] Banner will dismiss screen");
+//}
+//
+//- (void) adViewDidDismissScreen:(GADBannerView *)bannerView {
+//    NSLog(@"[SuperAwesome | AdMob] Banner did dismiss screen");
+//}
+//
+//- (void) adViewWillLeaveApplication:(GADBannerView *)bannerView {
+//    NSLog(@"[SuperAwesome | AdMob] Banner will leave application");
+//}
```
Comment (Tbruni85) [inline Example/SuperAwesome/AdMobController.m]: remove?

File: Example/SuperAwesome/AdMobController.m [hunk 8]
```diff
@@ -66,115 +66,115 @@ - (void)viewDidLoad {
...
+//    NSLog(@"[SuperAwesome | AdMob] Interstitial did dismiss screen");
+//}
+//
+//- (void) interstitialDidFailToPresentScreen:(GADInterstitial *)ad {
+//    NSLog(@"[SuperAwesome | AdMob] Interstitial did fail to present screen");
+//}
+//
+//- (void) interstitialWillLeaveApplication:(GADInterstitial *)ad {
+//    NSLog(@"[SuperAwesome | AdMob] Interatitial will leave application");
+//}
```
Comment (Tbruni85) [inline Example/SuperAwesome/AdMobController.m]: remove?

File: Example/SuperAwesome/AdMobController.m [hunk 9]
```diff
@@ -66,115 +66,115 @@ - (void)viewDidLoad {
...
+//
+//- (void)rewardBasedVideoAdWillLeaveApplication:(GADRewardBasedVideoAd *)rewardBasedVideoAd {
+//    NSLog(@"[SuperAwesome | AdMob] Video ad will leave application");
+//}
+//
+//- (void)rewardBasedVideoAd:(GADRewardBasedVideoAd *)rewardBasedVideoAd
+//   didRewardUserWithReward:(GADAdReward *)reward {
+//    NSLog(@"[SuperAwesome | AdMob] Video ad will be rewarded with %@ | %@", reward.amount, reward.type);
+//}
```
Comment (Tbruni85) [inline Example/SuperAwesome/AdMobController.m]: remove?

File: Example/SuperAwesome/AppDelegate.m [hunk 10]
```diff
@@ -8,7 +8,7 @@
 
 #import "AppDelegate.h"
 #import "AwesomeAds.h"
-#import <GoogleMobileAds/GoogleMobileAds.h>
+//#import <GoogleMobileAds/GoogleMobileAds.h>
```
Comment (Tbruni85) [inline Example/SuperAwesome/AppDelegate.m]: remove?

File: Example/SuperAwesome/AppDelegate.m [hunk 11]
```diff
@@ -20,7 +20,7 @@ @implementation AppDelegate
 - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
     // Override point for customization after application launch.
     
-    [GADMobileAds configureWithApplicationID:@"ca-app-pub-7706302691807937~3326320407"];
+//    [GADMobileAds configureWithApplicationID:@"ca-app-pub-7706302691807937~3326320407"];
```
Comment (Tbruni85) [inline Example/SuperAwesome/AppDelegate.m]: remove?

File: Example/SuperAwesome/ViewController.m [hunk 12]
```diff
@@ -25,8 +26,13 @@ @implementation ViewController
...
     
-    //    NSMutableArray *abc = @[@(13), @(25), @(88)];
-    //    [abc removeLastObject];
+//    [SAVideoAd2 setCallback:^(NSInteger placementId, SAEvent event) {
+//        if (event == adLoaded) {
+//            [SAVideoAd2 play:placementId fromVC:self];
+//        }
+//    }];
+//    [SAVideoAd2 setSmallClick:true];
+//    [SAVideoAd2 load:39521];
```
Comment (Tbruni85) [inline Example/SuperAwesome/ViewController.m]: remove?

File: Example/SuperAwesome/ViewController.m [hunk 13]
```diff
@@ -68,14 +74,13 @@ - (void)viewDidLoad {
     [SAVideoAd disableParentalGate];
     [SAVideoAd disableCloseButton];
     [SAVideoAd disableCloseAtEnd];
-    [SAVideoAd disableMoatLimiting];
+//    [SAVideoAd2 disableMoatLimiting];
```
Comment (Tbruni85) [inline Example/SuperAwesome/ViewController.m]: remove?

File: Pod/Classes/AwesomeAds.h [hunk 14]
```diff
@@ -6,7 +6,7 @@
 #import <UIKit/UIKit.h>
 #import "SABannerAd.h"
 #import "SAInterstitialAd.h"
-#import "SAVideoAd.h"
+//#import "SAVideoAd.h"
```
Comment (Tbruni85) [inline Pod/Classes/AwesomeAds.h]: remove?

File: Pod/Classes/CommonUI/LayoutUtils.swift [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    
+    static func bind(view: UIView, toTopRightOf otherView: UIView) {
+        view.translatesAutoresizingMaskIntoConstraints = false
+        let margins = otherView.layoutMarginsGuide
+        
+        view.topAnchor.constraint(equalTo: margins.topAnchor, constant: 0.0).isActive = true
+        view.trailingAnchor.constraint(equalTo: margins.trailingAnchor, constant: 0.0).isActive = true
+        view.widthAnchor.constraint(equalToConstant: 40.0).isActive = true
+        view.heightAnchor.constraint(equalToConstant: 40.0).isActive = true
+    }
```
Comment (Tbruni85) [inline Pod/Classes/CommonUI/LayoutUtils.swift]: nice done, **PureLayout** would have made you your life a bit easier tho.

File: Pod/Classes/SuperAwesome.h [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+//
+//  Created by Gabriel Coman on 17/12/2018.
+//
+
+//#ifndef SuperAwesome_h
+//#define SuperAwesome_h
+
+#import "SABannerAd.h"
+#import "SAInterstitialAd.h"
+//#import "SAVideoAd.h"
```
Comment (Tbruni85) [inline Pod/Classes/SuperAwesome.h]: remove?

File: Pod/Classes/SuperAwesome.h [hunk 17]
```diff
@@ -0,0 +1,27 @@
+//
+//  SuperAwesome.h
+//  SuperAwesome
+//
+//  Created by Gabriel Coman on 17/12/2018.
+//
+
+//#ifndef SuperAwesome_h
+//#define SuperAwesome_h
```
Comment (Tbruni85) [inline Pod/Classes/SuperAwesome.h]: remove?

File: Pod/Classes/Video/AdChromeControl.swift [hunk 18]
```diff
@@ -0,0 +1,176 @@
+//
+//  SAChromeControl.swift
```
Comment (Tbruni85) [inline Pod/Classes/Video/AdChromeControl.swift]: Chrome control is a very Android-ish naming, can we call it something more generic like `ControlView` or `PlayerControlView`?

File: Pod/Classes/Video/AdChromeControl.swift [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+            padlock.leadingAnchor.constraint(equalTo: margins.leadingAnchor, constant: 0.0).isActive = true
+            padlock.widthAnchor.constraint(equalToConstant: 67.0).isActive = true
+            padlock.heightAnchor.constraint(equalToConstant: 25.0).isActive = true
+        }
+        
+        closeButton = CloseButton()
+        closeButton.isHidden = !showCloseButton
+        closeButton.addTarget(self, action: #selector(close), for: .touchUpInside)
+        addSubview(closeButton)
+        LayoutUtils.bind(view: closeButton, toTopRightOf: self)
```
Comment (Tbruni85) [inline Pod/Classes/Video/AdChromeControl.swift]: to ensure the correct layout of the view you should add the constraint inside the 
`updateConstraints()` if it's a UIView
`updateViewConstraints()` if it's a UIViewController

In this case you'll need  a flag due to the fact those function are called, or might be called, multiple times and are not under your control.

You can see the way I implement them inside the UI SDK.

File: Pod/Classes/Video/AdChromeControl.swift [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+    }
+    
+    ////////////////////////////////////////////////////////////////////////////
+    // custom methods
+    ////////////////////////////////////////////////////////////////////////////
+    
+    @objc(makeCloseButtonVisible)
```
Comment (Tbruni85) [inline Pod/Classes/Video/AdChromeControl.swift]: isn't just `@objc` enough? the function name is the same

File: Pod/Plugin/MoPub/SAMoPubVideoCustomEvent.m [hunk 21]
```diff
@@ -73,74 +73,74 @@ - (void) requestRewardedVideoWithCustomEventInfo:(NSDictionary *)info {
...
+//            }
+//            case adClosed: {
+//                // call required events
+//                [weakSelf.delegate rewardedVideoWillDisappearForCustomEvent:weakSelf];
+//                [weakSelf.delegate rewardedVideoDidDisappearForCustomEvent:weakSelf];
+//                break;
+//            }
+//        }
+//    }];
+//    [SAVideoAd load:_placementId];
```
Comment (Tbruni85) [inline Pod/Plugin/MoPub/SAMoPubVideoCustomEvent.m]: remove?

File: Pod/Plugin/MoPub/SAMoPubVideoCustomEvent.m [hunk 22]
```diff
@@ -149,7 +149,7 @@ - (void) requestRewardedVideoWithCustomEventInfo:(NSDictionary *)info {
  * @param viewController the view controller from which the ad will spring
  */
 - (void) presentRewardedVideoFromViewController:(UIViewController *)viewController {
-    [SAVideoAd play:_placementId fromVC:viewController];
+//    [SAVideoAd play:_placementId fromVC:viewController];
```
Comment (Tbruni85) [inline Pod/Plugin/MoPub/SAMoPubVideoCustomEvent.m]: remove?

File: Example/SuperAwesome/AdMobController.m (hunk 1 above)
Comment (me2k) [reply Example/SuperAwesome/AdMobController.m]: I second that

File: Pod/Classes/CommonUI/LayoutUtils.swift (hunk 15 above)
Comment (me2k) [reply Pod/Classes/CommonUI/LayoutUtils.swift]: @devgabrielcoman we can help you make this pure layout if you wish? But it'll add another library to the project.

File: Pod/Classes/Video/AdChromeControl.swift (hunk 19 above)
Comment (me2k) [reply Pod/Classes/Video/AdChromeControl.swift]: This change is the correct way to add constraints, you will need to put them behind a flag `didSetUpConstraints` as they only need to be set up once unless you destroy them.

File: Pod/Classes/Video/VideoAd.swift [hunk 23]
```diff
@@ -0,0 +1,285 @@
...
+
+enum AdState {
+    case none
+    case loading
+    case hasAd(ad: SAAd)
+}
+
+@objc(SAVideoAd) public class VideoAd: NSObject {
+    
+    static var isTestingEnabled: Bool = Bool(truncating: NSNumber(value: SA_DEFAULT_TESTMODE))
```
Comment (me2k) [inline Pod/Classes/Video/VideoAd.swift]: Are these constants from the ObjC code? This is very strange to look at.

**Pull request merged**